massa-grpc/src/public.rs

```
@@ -0,0 +1,735 @@
+
+                            None
+                        })
+                        .collect::<Vec<Block>>();
+
+                    blocks.extend_from_slice(&result);
+                }
+            }
+        }
+        blocks
+    } else {
+        // only slot range is provided
+        let graph = grpc
+            .consensus_controller
+            .get_block_graph_status(slot_min.map(|s| s.into()), slot_max.map(|s| s.into()))?;
+
+        graph
+            .active_blocks
+            .iter()
+            .filter_map(|b| {
+                read_blocks
+                    .get(b.0)
+                    .map(|wrapped_block| wrapped_block.content.clone())
+            })
+            .collect::<Vec<Block>>()
+    };
+
+    let blocks_ids = blocks
+        .iter()
+        .map(|block| block.header.id)
+        .collect::<Vec<BlockId>>();
+
+    let blocks_status = grpc.consensus_controller.get_block_statuses(&blocks_ids);
+
+    let result = blocks
+        .iter()
+        .zip(blocks_status)
+        .map(|(block, block_graph_status)| grpc_model::BlockWrapper {
+            block_id: block.header.id.to_string(),
+            block: Some(block.clone().into()),
+            status: block_graph_status.into(),
+        })
+        .collect();
+
+    Ok(grpc_api::GetBlocksResponse {
+        wrapped_blocks: result,
+    })
+}
+
+/// Get multiple datastore entries
+pub(crate) fn get_datastore_entries(
+    grpc: &MassaPublicGrpc,
+    request: tonic::Request<grpc_api::GetDatastoreEntriesRequest>,
+) -> Result<grpc_api::GetDatastoreEntriesResponse, GrpcError> {
+    let inner_req = request.into_inner();
+
+    let filters: Vec<(Address, Vec<u8>)> = inner_req
+        .filters
+        .into_iter()
+        .filter_map(|filter| {
+            filter.filter.and_then(|filter| match filter {
+                grpc_api::get_datastore_entry_filter::Filter::AddressKey(addrs) => {
+                    if let Ok(add) = &Address::from_str(&addrs.address) {
```

@modship we should fail fast no ?